server/src/main/java/org/opensearch/index/store/remote/file/OnDemandCompositeBlockIndexInput.java

```diff
@@ -0,0 +1,229 @@
+
+    public OnDemandCompositeBlockIndexInput(
+        RemoteSegmentStoreDirectory remoteDirectory,
+        String fileName,
+        FSDirectory localDirectory,
+        FileCache fileCache,
+        IOContext context
+    ) throws IOException {
+        this(
+            OnDemandBlockIndexInput.builder()
+                .resourceDescription("OnDemandCompositeBlockIndexInput")
+                .isClone(false)
+                .offset(0L)
+                .length(remoteDirectory.fileLength(fileName)),
+            remoteDirectory,
+            fileName,
+            localDirectory,
+            fileCache,
+            context
+        );
+    }
+
+    public OnDemandCompositeBlockIndexInput(
+        Builder builder,
+        RemoteSegmentStoreDirectory remoteDirectory,
+        String fileName,
+        FSDirectory localDirectory,
+        FileCache fileCache,
+        IOContext context
+    ) throws IOException {
+        super(builder);
+        this.remoteDirectory = remoteDirectory;
+        this.localDirectory = localDirectory;
+        this.fileName = fileName;
+        this.fileCache = fileCache;
+        this.context = context;
+        originalFileSize = remoteDirectory.fileLength(fileName);
+    }
+
+    @Override
+    protected OnDemandCompositeBlockIndexInput buildSlice(String sliceDescription, long offset, long length) {
+        try {
+            return new OnDemandCompositeBlockIndexInput(
+                OnDemandBlockIndexInput.builder()
+                    .blockSizeShift(blockSizeShift)
+                    .isClone(true)
+                    .offset(this.offset + offset)
+                    .length(length)
+                    .resourceDescription(sliceDescription),
+                remoteDirectory,
+                fileName,
+                localDirectory,
+                fileCache,
+                context
+            );
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    protected IndexInput fetchBlock(int blockId) throws IOException {
+        logger.trace("fetchBlock called with blockId -> {}", blockId);
+        final String blockFileName = fileName + "_block_" + blockId;
+        final long blockStart = getBlockStart(blockId);
+        final long length = getActualBlockSize(blockId);
+        logger.trace(
+            "File: {} , Block File: {} , Length: {} , BlockSize: {} , OriginalFileSize: {}",
+            fileName,
+            blockFileName,
+            blockStart,
+            length,
+            originalFileSize
+        );
+        Path blockFilePath = getLocalFilePath(blockFileName);
+        final CachedIndexInput cacheEntry = fileCache.compute(blockFilePath, (path, cachedIndexInput) -> {
+            if (cachedIndexInput == null || cachedIndexInput.isClosed()) {
+                // Doesn't exist or is closed, either way create a new one
+                IndexInput indexInput = fetchIndexInput(blockFileName, blockStart, length);
```

Comment:
See #7031 for more details, but you don't want to do the blocking work inside `FileCache#compute`. The cache implementation is ultimately pretty simple and once you get down to a cache "segment" then thread safety is provided with [simple locks](https://github.com/opensearch-project/OpenSearch/blob/a1e42b1131d5d807dc66d13dd960ebe707da3f09/server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java#L148-L173). The upshot is that this implementation will result in completely unrelated blocks (that just so happen to land on the same cache segment) blocking on one another.

This [TransferManager class](https://github.com/opensearch-project/OpenSearch/blob/3125b948029609f354d3153f8ca6391638daefc7/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java#L39) (which is a terribly generic and unhelpful name) implements the pattern that defers the blocking work (remote and file I/O) outside the critical section of the cache. It seems like the hang up here for reusing that code is that you're using a `Directory` to do the remote I/O while the existing code is using a `BlobContainer`. Would definitely be good to see if we can bridge that gap and reuse code where possible (similar to @mch2's comment).

Reply:
You'll also want to look at #6641. Weird things happen with regard to permissions when the search is triggered from a scripted query and that search request ends up needing to do remote or local file I/O. If you can find a way to refactor/reuse the existing implementation in TransferManager you probably won't need to worry about this. If not then you'll have to wrap the requisite calls in `AccessController#doPrivileged`.

Reply:
Makes sense to use the Transfer Manager instead of duplicating it.
To do so have sort out abstracted out the logic that will be generic (such as creating a DelayedCreationIndexInput and making I/O calls inside AccessController#doPrivileged.
We can create a concrete implementation of the abstraction created based on the logic that will be used to serve the I/O request - whether from BlobContainer or from Remote Directory.

[AbstractTransferManager](https://github.com/opensearch-project/OpenSearch/blob/70bac1fd8bb04337bb26f738ddb620bb63322f86/server/src/main/java/org/opensearch/index/store/remote/utils/AbstractTransferManager.java) -> containing the common logic

[TransferManager](https://github.com/opensearch-project/OpenSearch/blob/70bac1fd8bb04337bb26f738ddb620bb63322f86/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java) -> specific to RemoteSnapshot use case where we fetch from BlobContainer

[CompositeDirectoryTransferManager](https://github.com/opensearch-project/OpenSearch/blob/70bac1fd8bb04337bb26f738ddb620bb63322f86/server/src/main/java/org/opensearch/index/store/remote/utils/CompositeDirectoryTransferManager.java) -> specific to Composite Directory where we fetch from RemoteDirectory

Does the above look fine to you ?

-----



> This [TransferManager class](https://github.com/opensearch-project/OpenSearch/blob/3125b948029609f354d3153f8ca6391638daefc7/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java#L39) (which is a terribly generic and unhelpful name)

Agree on this part. Currently not refactoring it to avoid touching the Remote Snapshot flow in this PR. Will raise a separate one to refactor it(along with some other refactoring that might be needed), if it is ok with you.

Reply:
Changed the implementation of Transfer Manager to include a functional interface for reading the stream